bluefog/common/mpi_controller.cc

```diff
@@ -432,10 +470,75 @@ void MPIController::NeighborAllreduce(TensorTableEntry& entry) {
             "output for details.");
       }
     } else {
-      error_message = mpi_ctx_.NeighborValueExchangeWithConstantElements(
-        sendbuf, (void *)entry.output->data(), num_elements, entry.output->dtype(),
-        entry.send_neighbors.get(), entry.recv_neighbors.get()
-      );
+      int nsend = entry.send_neighbors->size();
+      int nrecv = entry.recv_neighbors->size();
+
+      // Ensure the lifecycle of the weighted tensors are alive after communication.
+      std::vector<std::unique_ptr<common::Tensor>> weighted_tensors;
+      if (entry.dst_weighting_enabled) {
+        for (int i = 0; i < nsend; ++i) {
+          auto weighted_tensor_ptr = entry.tensor->data_weight(entry.send_weights->at(i));
+          weighted_tensors.push_back(std::move(weighted_tensor_ptr));
+        }
+      }
+      std::shared_ptr<common::ReadyEvent> ready_event =
+          entry.context->RecordReadyEvent(entry.device);
```

Comment:
Instead of call the Ready function externally, we should have API like `tensor->data_weight_blocking(w)`. The external should not care how and when to make it ready?

Reply:
Add #83 and TODO.